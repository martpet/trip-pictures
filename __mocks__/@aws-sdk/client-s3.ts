export const S3Client = function DummyS3Client() {};
