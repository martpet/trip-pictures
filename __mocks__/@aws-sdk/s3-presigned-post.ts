export const createPresignedPost = vi
  .fn()
  .mockName('createPresignedPost')
  .mockResolvedValue('dummyPresignedPost');
