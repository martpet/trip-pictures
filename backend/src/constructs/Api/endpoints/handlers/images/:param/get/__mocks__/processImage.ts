export const processImage = vi
  .fn()
  .mockName('processImage')
  .mockResolvedValue('dummyBase64EncodedImage');
