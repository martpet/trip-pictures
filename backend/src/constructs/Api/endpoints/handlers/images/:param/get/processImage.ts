import { SdkStreamMixin } from '@aws-sdk/types';
import { DetectModerationLabelsCommand, RekognitionClient } from 'lambda-layer';
import sharp from 'sharp';

const rekognition = new RekognitionClient({});

type Props = {
  stream: SdkStreamMixin;
  quality?: number;
  width?: number;
  height?: number;
};

export async function processImage({
  stream,
  width = 1500,
  height,
  quality = 100,
}: Props) {
  const uint8 = await stream.transformToByteArray();
  const image = sharp(uint8).rotate().resize(width, height).jpeg({ quality });
  let buffer = await image.toBuffer();

  const moderationCommand = new DetectModerationLabelsCommand({
    Image: { Bytes: buffer },
    MinConfidence: 90,
  });

  const { ModerationLabels } = await rekognition.send(moderationCommand);

  const isModerated = ModerationLabels?.some(({ Name }) => Name === 'Explicit Nudity');

  if (isModerated) {
    image.blur(75);
    buffer = await image.toBuffer();
  }

  const base64 = buffer.toString('base64');
  return base64;
}
