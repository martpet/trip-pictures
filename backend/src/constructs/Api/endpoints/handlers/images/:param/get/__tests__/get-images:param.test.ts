import { GetObjectCommand, GetObjectCommandOutput, S3Client } from '@aws-sdk/client-s3';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import { lambdaPayloadLimit } from 'lambda-layer';

import {
  itCalls,
  itHasEnvVars,
  itHasPathParam,
  itResolves,
  itResolvesWithError,
  itSendsAwsCommand,
} from '~/constructs/Api/utils';

import { handler } from '../get-images:param';
import { processImage } from '../processImage';

vi.mock('../processImage');
vi.mock('~/constructs/Api/utils/errorResponse');

const ddbMock = mockClient(DynamoDBDocumentClient);
const s3Mock = mockClient(S3Client);

process.env.photoBucket = 'dummyPhotoBucketName';

const args = [
  {
    pathParameters: {
      fingerprint: 'dummyFingerprint',
    },
    queryStringParameters: {
      quality: '99',
      width: '888',
      height: '555',
    },
  },
] as unknown as Parameters<typeof handler>;

beforeEach(() => {
  ddbMock.reset();
  s3Mock.reset();

  ddbMock.on(GetCommand).resolves({
    Item: { userId: 'dummyUserId' },
  });

  s3Mock.on(GetObjectCommand).resolves({
    Body: 'dummyBody',
  } as unknown as GetObjectCommandOutput);
});

describe('get-images:param', () => {
  itHasEnvVars(['photoBucket'], handler, args);
  itHasPathParam('fingerprint', handler, args);
  itSendsAwsCommand(GetCommand, ddbMock, handler, args);
  itSendsAwsCommand(GetObjectCommand, s3Mock, handler, args);
  itCalls(processImage, handler, args);
  itResolves(handler, args);

  describe.each(['width', 'height', 'quality'])(
    'when "%s" query string parameter is missing',
    (key) => {
      const argsClone = structuredClone(args);
      const queryStrings = { ...args[0].queryStringParameters };
      delete queryStrings[key];
      argsClone[0].queryStringParameters = queryStrings;
      itCalls(processImage, handler, argsClone);
    }
  );

  describe('when "Item" is missing from "GetCommand" output', () => {
    beforeEach(() => {
      ddbMock.on(GetCommand).resolves({});
    });
    itResolvesWithError(handler, args);
  });

  describe('when "Body" is missing from "GetObjectCommand" output', () => {
    beforeEach(() => {
      s3Mock.on(GetObjectCommand).resolves({});
    });
    itResolvesWithError(handler, args);
  });

  describe('when "processImage" output length is bigger than lambda payload limit', () => {
    beforeEach(() => {
      vi.mocked(processImage).mockResolvedValue({
        length: lambdaPayloadLimit + 1,
      } as any);
    });
    itResolvesWithError(handler, args);
  });
});
