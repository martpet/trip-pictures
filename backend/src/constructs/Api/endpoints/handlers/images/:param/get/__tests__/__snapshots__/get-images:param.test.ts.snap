// Vitest Snapshot v1

exports[`get-images:param > calls "processImage" with correct args 1`] = `
[
  [
    {
      "height": 555,
      "quality": 99,
      "stream": "dummyBody",
      "width": 888,
    },
  ],
]
`;

exports[`get-images:param > resolves with a correct value 1`] = `
{
  "body": "dummyBase64EncodedImage",
  "headers": {
    "Cache-Control": "max-age=3153600000",
    "Content-Type": "image/jpeg",
  },
  "isBase64Encoded": true,
  "statusCode": 200,
}
`;

exports[`get-images:param > sends "GetCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "Key": {
    "fingerprint": "dummyFingerprint",
  },
  "TableName": "TP-photos",
}
`;

exports[`get-images:param > sends "GetObjectCommand" from "S3Client" with correct args 1`] = `
{
  "Bucket": "dummyPhotoBucketName",
  "Key": "dummyUserId/dummyFingerprint",
}
`;

exports[`get-images:param > when "Body" is missing from "GetObjectCommand" output > calls "errorResponse" with correct args 1`] = `
[
  [
    "dc1c2c42bf",
  ],
]
`;

exports[`get-images:param > when "Body" is missing from "GetObjectCommand" output > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images:param > when "Item" is missing from "GetCommand" output > calls "errorResponse" with correct args 1`] = `
[
  [
    "5b2797f93a",
    {
      "statusCode": 404,
    },
  ],
]
`;

exports[`get-images:param > when "Item" is missing from "GetCommand" output > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images:param > when "fingerprint" path parameter is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "564208f603",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`get-images:param > when "fingerprint" path parameter is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images:param > when "height" query string parameter is missing > calls "processImage" with correct args 1`] = `
[
  [
    {
      "height": undefined,
      "quality": 99,
      "stream": "dummyBody",
      "width": 888,
    },
  ],
]
`;

exports[`get-images:param > when "photoBucket" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "68e8d0232f",
  ],
]
`;

exports[`get-images:param > when "photoBucket" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images:param > when "processImage" output length is bigger than lambda payload limit > calls "errorResponse" with correct args 1`] = `
[
  [
    "5fe36afa17",
  ],
]
`;

exports[`get-images:param > when "processImage" output length is bigger than lambda payload limit > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`get-images:param > when "quality" query string parameter is missing > calls "processImage" with correct args 1`] = `
[
  [
    {
      "height": 555,
      "quality": undefined,
      "stream": "dummyBody",
      "width": 888,
    },
  ],
]
`;

exports[`get-images:param > when "width" query string parameter is missing > calls "processImage" with correct args 1`] = `
[
  [
    {
      "height": 555,
      "quality": 99,
      "stream": "dummyBody",
      "width": undefined,
    },
  ],
]
`;
